src/devices/Sn74hc595/Sn74hc595.cs

```diff
@@ -0,0 +1,354 @@
+            }
+
+            _controller.Write(_data, value);
+            _controller.Write(_srclk, 1);
+            _controller.Write(_data, 0);
+            _controller.Write(_srclk, 0);
+        }
+
+        /// <summary>
+        /// Shifts a byte -- 8 bits -- to the 8 registers.
+        /// Pushes / overwrites any existing values.
+        /// Latches by default.
+        /// </summary>
+        public void ShiftByte(byte value, bool latch = true)
+        {
+            if (_spiDevice is object)
+            {
+                _spiDevice.WriteByte(value);
+                return;
+            }
+
+            for (int i = 0; i < 8; i++)
+            {
+                int data = (128 >> i) & value;
+                ShiftBit(data);
+            }
+
+            if (latch)
+            {
+                Latch();
+            }
+        }
+
+        /// <summary>
+        /// Latches values in data register.
+        /// Essentially a "publish" command.
+        /// Requires use of GPIO controller.
+        /// </summary>
+        public void Latch()
+        {
+            if (_controller is null || _pinMapping.RClk == 0)
+            {
+                throw new ArgumentNullException($"{nameof(ShiftBit)}: GpioController was not provided or {nameof(_pinMapping.RClk)} not mapped to pin");
+            }
+
+            _controller.Write(_rclk, 1);
+            _controller.Write(_rclk, 0);
+        }
+
+        /// <summary>
+        /// Switch output register to high-impedance state.
+        /// Disables current output register.
+        /// Requires use of GPIO controller.
+        /// </summary>
+        public void OutputDisable()
+        {
+            if (_controller is null || _pinMapping.OE == 0)
+            {
+                throw new ArgumentNullException($"{nameof(OutputDisable)}: {nameof(_pinMapping.OE)} not mapped to non-zero pin value");
+            }
+
+            _controller.Write(_pinMapping.OE, 1);
+        }
+
+        /// <summary>
+        /// Switch output register low-impedance state.
+        /// Enables current register.
+        /// Requires use of GPIO controller.
+        /// </summary>
+        public void OutputEnable()
+        {
+            if (_controller is null || _pinMapping.OE == 0)
+            {
+                throw new ArgumentNullException($"{nameof(OutputEnable)}: {nameof(_pinMapping.OE)} not mapped to non-zero pin value");
+            }
+
+            _controller.Write(_pinMapping.OE, 0);
+        }
+
+        /// <summary>
+        /// Cleanup.
+        /// Failing to dispose this class, especially when callbacks are active, may lead to undefined behavior.
+        /// </summary>
+        public void Dispose()
+        {
+            if (_shouldDispose)
+            {
+                _controller?.Dispose();
+                _controller = null;
+                _spiDevice?.Dispose();
```

Comment:
We usually always dispose spidevices given that in most cases they will only serve one single device whereas a gpiocontroller can be used outside. All that said, given we have the shouldDispose what you are doing is probably the right thing to do.

Reply:
I could go either way. I took your feedback and moved the SPI logic out of the condition.